src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -905,29 +905,27 @@ namespace cryptonote
       return 1;
     }
 
-    std::vector<cryptonote::blobdata> newtxs;
-    newtxs.reserve(arg.txs.size());
+    std::vector<tx_blob_entry> entries;
+    entries.reserve(arg.txs.size());
     for (size_t i = 0; i < arg.txs.size(); ++i)
     {
-      cryptonote::tx_verification_context tvc = AUTO_VAL_INIT(tvc);
-      m_core.handle_incoming_tx({arg.txs[i], crypto::null_hash}, tvc, false, true, false);
-      if(tvc.m_verifivation_failed)
+      entries.push_back({arg.txs[i], crypto::null_hash});
+    }
+    std::vector<cryptonote::tx_verification_context> tvc;
+    epee::net_utils::network_address address = context.m_remote_address;
+    boost::uuids::uuid connection_id = context.m_connection_id;
+    m_core.handle_incoming_txs(entries, tvc, [address, connection_id, this](bool valid) {
+      MDEBUG("tx verified callback: address " << address.str() << ", connection " << connection_id << ": " << (valid ? "valid" : "invalid"));
+      if (!valid)
       {
-        LOG_PRINT_CCONTEXT_L1("Tx verification failed, dropping connection");
-        drop_connection(context, false, false);
-        return 1;
+        // drop connection (this will bump the ban score), and manually ban if the connection doesn't exist anymore
+        if (!m_p2p->for_connection(connection_id, [&](cryptonote_connection_context& context, nodetool::peerid_type peer_id, uint32_t f)->bool {
```

Comment:
Shouldn't we always drop the connection and bump the ban score. I.e. without `if`

Reply:
This means: if we could not find the connection (and ban it), ban it manually. This is because since verification is now defered, the connection might have already been dropped for other reasons, and the peer would not get banned, so it could loop connect/send-bad-tx/disconnect.

Reply:
Yep, didn't see that `drop_connection` invokes `add_host_fail` internally. Looks okay